wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/FederatedAuthConnectionPlugin.java

```diff
@@ -0,0 +1,347 @@
+    try {
+      Class.forName("software.amazon.awssdk.services.sts.model.AssumeRoleWithSamlRequest");
+    } catch (final ClassNotFoundException e) {
+      throw new RuntimeException(Messages.get("FederatedAuthConnectionPlugin.javaStsSdkNotInClasspath"));
+    }
+    this.pluginService = pluginService;
+    this.credentialsProviderFactory = credentialsProviderFactory;
+    this.telemetryFactory = pluginService.getTelemetryFactory();
+    this.cacheSizeGauge = telemetryFactory.createGauge("federatedAuth.tokenCache.size", () -> (long) tokenCache.size());
+    this.fetchTokenCounter = telemetryFactory.createCounter("federatedAuth.fetchToken.count");
+  }
+
+
+  @Override
+  public Connection connect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, connectFunc);
+  }
+
+  @Override
+  public Connection forceConnect(
+      final @NonNull String driverProtocol,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties props,
+      final boolean isInitialConnection,
+      final @NonNull JdbcCallable<Connection, SQLException> forceConnectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, forceConnectFunc);
+  }
+
+  private Connection connectInternal(HostSpec hostSpec, Properties props,
+      JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+
+    String host = getHost(hostSpec, props);
+
+    int port = getPort(hostSpec, props);
+
+    final Region region = getRegion(host, props);
+
+    final String cacheKey = getCacheKey(
+        PropertyDefinition.USER.getString(props),
+        host,
+        port,
+        region);
+
+    final TokenInfo tokenInfo = tokenCache.get(cacheKey);
+
+    final boolean isCachedToken = tokenInfo != null && !tokenInfo.isExpired();
+
+    if (isCachedToken) {
+      LOGGER.finest(
+          () -> Messages.get(
+              "FederatedAuthConnectionPlugin.useCachedIamToken",
+              new Object[] {tokenInfo.getToken()}));
+      PropertyDefinition.PASSWORD.set(props, tokenInfo.getToken());
+    } else {
+      getSetAndCacheAuthenticationToken(hostSpec, props, region, cacheKey);
+    }
+
+    try {
+      return connectFunc.call();
+    } catch (final SQLException exception) {
+      getSetAndCacheAuthenticationToken(hostSpec, props, region, cacheKey);
+      return connectFunc.call();
+    } catch (final Exception exception) {
+      LOGGER.warning(
+          () -> Messages.get(
+              "FederatedAuthConnectionPlugin.unhandledException",
+              new Object[] {exception}));
+      throw new SQLException(exception);
+    }
+  }
+
+  private void getSetAndCacheAuthenticationToken(HostSpec hostSpec, Properties props, Region region, String cacheKey)
```

Comment:
Usually methods with names like `getXXX` are expected to return some value. Would you consider to rename it?

Reply:
fetch and store could be update() ?